[TensorExpr] Add a fuser pass based on tensor expressions.

ZolotukhinM opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #34234 [WIP] Changes in test/test_jit_fuser.py test/test_jit_fuser_te.py.
* #34233 [WIP] Changes to peephole.cpp.
* #34231 [WIP] Guard elimination changes from bertmaher/pytorch_fusion.
* #34230 [TensorExpr] Add tensorexpr benchmarks.
* #34228 [TensorExpr] Add LLVM codegen.
* #34227 [TensorExpr] Add CUDA codegen.
* **#34226 [TensorExpr] Add a fuser pass based on tensor expressions.**
* #34224 [TensorExpr] Pull changes to core classes for representing expressions and statements from the side branch.

LLVM and Cuda backends are added in subsequent PRs, so at this point the fuser is pretty useless, but it still can be tested and its logic is not going to change with addition of the codegens.

Differential Revision: [D20251838](https://our.internmc.facebook.com/intern/diff/D20251838)

Commit:
Tensorexpr fuser implementation.

[ghstack-poisoned]

Commit:
Update on "Tensorexpr fuser implementation."


Differential Revision: [D20251838](https://our.internmc.facebook.com/intern/diff/D20251838)

[ghstack-poisoned]

Commit:
Update descriptions. on "[TensorExpr] Add a fuser pass based on tensor expressions."


LLVM and Cuda backends are added in subsequent PRs, so at this point the fuser is pretty useless, but it still can be tested and its logic is not going to change with addition of the codegens.

Differential Revision: [D20251838](https://our.internmc.facebook.com/intern/diff/D20251838)

[ghstack-poisoned]

Commit:
Lints on "[TensorExpr] Add a fuser pass based on tensor expressions."


LLVM and Cuda backends are added in subsequent PRs, so at this point the fuser is pretty useless, but it still can be tested and its logic is not going to change with addition of the codegens.

Differential Revision: [D20251838](https://our.internmc.facebook.com/intern/diff/D20251838)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add a fuser pass based on tensor expressions."


LLVM and Cuda backends are added in subsequent PRs, so at this point the fuser is pretty useless, but it still can be tested and its logic is not going to change with addition of the codegens.

Differential Revision: [D20251838](https://our.internmc.facebook.com/intern/diff/D20251838)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add a fuser pass based on tensor expressions."


LLVM and Cuda backends are added in subsequent PRs, so at this point the fuser is pretty useless, but it still can be tested and its logic is not going to change with addition of the codegens.

Differential Revision: [D20251838](https://our.internmc.facebook.com/intern/diff/D20251838)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add a fuser pass based on tensor expressions."


LLVM and Cuda backends are added in subsequent PRs, so at this point the fuser is pretty useless, but it still can be tested and its logic is not going to change with addition of the codegens.

Differential Revision: [D20251838](https://our.internmc.facebook.com/intern/diff/D20251838)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add a fuser pass based on tensor expressions."


LLVM and Cuda backends are added in subsequent PRs, so at this point the fuser is pretty useless, but it still can be tested and its logic is not going to change with addition of the codegens.

Differential Revision: [D20251838](https://our.internmc.facebook.com/intern/diff/D20251838)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add a fuser pass based on tensor expressions."


LLVM and Cuda backends are added in subsequent PRs, so at this point the fuser is pretty useless, but it still can be tested and its logic is not going to change with addition of the codegens.

Differential Revision: [D20251838](https://our.internmc.facebook.com/intern/diff/D20251838)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add a fuser pass based on tensor expressions."


LLVM and Cuda backends are added in subsequent PRs, so at this point the fuser is pretty useless, but it still can be tested and its logic is not going to change with addition of the codegens.

Differential Revision: [D20251838](https://our.internmc.facebook.com/intern/diff/D20251838)

[ghstack-poisoned]

Commit:
Rebase on "[TensorExpr] Add a fuser pass based on tensor expressions."


LLVM and Cuda backends are added in subsequent PRs, so at this point the fuser is pretty useless, but it still can be tested and its logic is not going to change with addition of the codegens.

Differential Revision: [D20251838](https://our.internmc.facebook.com/intern/diff/D20251838)

[ghstack-poisoned]

Commit:
Rebase on "[TensorExpr] Add a fuser pass based on tensor expressions."


LLVM and Cuda backends are added in subsequent PRs, so at this point the fuser is pretty useless, but it still can be tested and its logic is not going to change with addition of the codegens.

Differential Revision: [D20251838](https://our.internmc.facebook.com/intern/diff/D20251838)

[ghstack-poisoned]

Commit:
Rebase on "[TensorExpr] Add a fuser pass based on tensor expressions."


LLVM and Cuda backends are added in subsequent PRs, so at this point the fuser is pretty useless, but it still can be tested and its logic is not going to change with addition of the codegens.

Differential Revision: [D20251838](https://our.internmc.facebook.com/intern/diff/D20251838)

[ghstack-poisoned]

bertmaher reviewed (approved):
Reviewed in side branch: http://github.com/bertmaher/pytorch/tree/pytorch_fusion/

Commit:
Rebase and try to fix windows build. on "[TensorExpr] Add a fuser pass based on tensor expressions."


LLVM and Cuda backends are added in subsequent PRs, so at this point the fuser is pretty useless, but it still can be tested and its logic is not going to change with addition of the codegens.

Differential Revision: [D20251838](https://our.internmc.facebook.com/intern/diff/D20251838)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add a fuser pass based on tensor expressions."


LLVM and Cuda backends are added in subsequent PRs, so at this point the fuser is pretty useless, but it still can be tested and its logic is not going to change with addition of the codegens.

Differential Revision: [D20251838](https://our.internmc.facebook.com/intern/diff/D20251838)

[ghstack-poisoned]

suo reviewed (approved):
lg from integration perspective.

Commit:
Update on "[TensorExpr] Add a fuser pass based on tensor expressions."


LLVM and Cuda backends are added in subsequent PRs, so at this point the fuser is pretty useless, but it still can be tested and its logic is not going to change with addition of the codegens.

Differential Revision: [D20251838](https://our.internmc.facebook.com/intern/diff/D20251838)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add a fuser pass based on tensor expressions."


LLVM and Cuda backends are added in subsequent PRs, so at this point the fuser is pretty useless, but it still can be tested and its logic is not going to change with addition of the codegens.

Differential Revision: [D20251838](https://our.internmc.facebook.com/intern/diff/D20251838)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add a fuser pass based on tensor expressions."


LLVM and Cuda backends are added in subsequent PRs, so at this point the fuser is pretty useless, but it still can be tested and its logic is not going to change with addition of the codegens.

Differential Revision: [D20251838](https://our.internmc.facebook.com/intern/diff/D20251838)

[ghstack-poisoned]

Commit:
Rebase on "[TensorExpr] Add a fuser pass based on tensor expressions."


LLVM and Cuda backends are added in subsequent PRs, so at this point the fuser is pretty useless, but it still can be tested and its logic is not going to change with addition of the codegens.

Differential Revision: [D20251838](https://our.internmc.facebook.com/intern/diff/D20251838)

[ghstack-poisoned]
